File: gotOops.cpp
#include <iostream>
#include <memory>
using namespace std;

// === Strategy Interface ===
class BattleCryStrategy {
public:
    virtual void battleCry() const = 0;
    virtual ~BattleCryStrategy() = default;
};

// === Concrete Strategies ===
class DefaultBattleCry : public BattleCryStrategy {
public:
    void battleCry() const override {
        cout << "For Westeros!" << endl;
    }
};

class StarkBattleCry : public BattleCryStrategy {
public:
    void battleCry() const override {
        cout << "The North Remembers!" << endl;
    }
};

class TargaryenBattleCry : public BattleCryStrategy {
public:
    void battleCry() const override {
        cout << "Dracarys!" << endl;
    }
};

class LannisterBattleCry : public BattleCryStrategy {
public:
    void battleCry() const override {
        cout << "Hear Me Roar!" << endl;
    }
};

// === Base Character Class ===
class Character {
protected:
    string name;
    string house;
    string realParents = "Unknown";
    unique_ptr<BattleCryStrategy> battleCryStrategy;

public:
    Character(string n, string h, BattleCryStrategy* strategy)
        : name(move(n)), house(move(h)), battleCryStrategy(strategy) {}

    virtual ~Character() = default;

    virtual void introduce() const {
        cout << "I am " << name << " of House " << house << "." << endl;
    }

    virtual void fight() const {
        cout << name << " swings a sword!" << endl;
    }

    virtual void battleCry() const {
        battleCryStrategy->battleCry();
    }

    void setRealParents(const string& rp) {
        realParents = rp;
    }

    string revealParents() const {
        return "My real parents are: " + realParents;
    }

    string getName() const { return name; }
};

// === Concrete Character Classes ===
class Stark : public Character {
public:
    Stark(string name)
        : Character(move(name), "Stark", new StarkBattleCry()) {}

    void callDirewolf() const {
        cout << name << " whistles. A direwolf appears!" << endl;
    }
};

class Targaryen : public Character {
public:
    Targaryen(string name)
        : Character(move(name), "Targaryen", new TargaryenBattleCry()) {}

    void summonDragon() const {
        cout << name << " shouts. A dragon lands nearby." << endl;
    }
};

class Lannister : public Character {
public:
    Lannister(string name)
        : Character(move(name), "Lannister", new LannisterBattleCry()) {}

    void payDebt() const {
        cout << name << " pays their debts (or sends someone to do it)." << endl;
    }
};

// === Main ===
int main() {
    Stark jon("Jon Snow");
    Targaryen dany("Daenerys Targaryen");
    Lannister tyrion("Tyrion Lannister");

    jon.setRealParents("Rhaegar Targaryen & Lyanna Stark");

    jon.introduce();
    dany.introduce();
    tyrion.introduce();

    cout << "\n--- Battle Cries ---" << endl;
    jon.battleCry();
    dany.battleCry();
    tyrion.battleCry();

    cout << "\n--- Special Abilities ---" << endl;
    jon.callDirewolf();
    dany.summonDragon();
    tyrion.payDebt();

    cout << "\n--- Fights ---" << endl;
    jon.fight();
    dany.fight();
    tyrion.fight();

    cout << "\n--- Family Secrets ---" << endl;
    cout << jon.revealParents() << endl;

    return 0;
}

File: GotOOPs.java
interface BattleCryStrategy {
    void battleCry();
}

class DefaultBattleCry implements BattleCryStrategy {
    public void battleCry() {
        System.out.println("For Westeros!");
    }
}

class StarkBattleCry implements BattleCryStrategy {
    public void battleCry() {
        System.out.println("The North Remembers!");
    }
}

class TargaryenBattleCry implements BattleCryStrategy {
    public void battleCry() {
        System.out.println("Dracarys!");
    }
}

class LannisterBattleCry implements BattleCryStrategy {
    public void battleCry() {
        System.out.println("Hear Me Roar!");
    }
}

class Character {
    private String name;
    private String house;
    private String realParents = "Unknown";
    private BattleCryStrategy battleCryStrategy;

    public Character(String name, String house, BattleCryStrategy strategy) {
        this.name = name;
        this.house = house;
        this.battleCryStrategy = strategy;
    }

    public void introduce() {
        System.out.println("I am " + name + " of House " + house + ".");
    }

    public void fight() {
        System.out.println(name + " swings a sword!");
    }

    public void battleCry() {
        battleCryStrategy.battleCry();
    }

    public void setRealParents(String realParents) {
        this.realParents = realParents;
    }

    public String revealParents() {
        return "My real parents are: " + realParents;
    }

    public String getName() {
        return name;
    }
}

class Stark extends Character {
    public Stark(String name) {
        super(name, "Stark", new StarkBattleCry());
    }

    public void callDirewolf() {
        System.out.println(getName() + " whistles. A direwolf appears!");
    }
}

class Targaryen extends Character {
    public Targaryen(String name) {
        super(name, "Targaryen", new TargaryenBattleCry());
    }

    public void summonDragon() {
        System.out.println(getName() + " shouts. A dragon lands nearby.");
    }
}

class Lannister extends Character {
    public Lannister(String name) {
        super(name, "Lannister", new LannisterBattleCry());
    }

    public void payDebt() {
        System.out.println(getName() + " pays their debts (or sends someone to do it).");
    }
}

public class Main {
    public static void main(String[] args) {
        Stark jon = new Stark("Jon Snow");
        Targaryen dany = new Targaryen("Daenerys Targaryen");
        Lannister tyrion = new Lannister("Tyrion Lannister");

        jon.setRealParents("Rhaegar Targaryen & Lyanna Stark");

        jon.introduce();
        dany.introduce();
        tyrion.introduce();

        System.out.println("\n--- Battle Cries ---");
        jon.battleCry();
        dany.battleCry();
        tyrion.battleCry();

        System.out.println("\n--- Special Abilities ---");
        jon.callDirewolf();
        dany.summonDragon();
        tyrion.payDebt();

        System.out.println("\n--- Fights ---");
        jon.fight();
        dany.fight();
        tyrion.fight();

        System.out.println("\n--- Family Secrets ---");
        System.out.println(jon.revealParents());
    }
}
